### ISSUE TITLE
Unable to initialize table 'security'

### ISSUE THREAD

**Issue by seblu** [firewalld/firewalld#484]
Firewalld displays error when starting with kernel without security tables. 

```
# LANG=C.UTF8 journalctl --no-host --utc --no-pager -u firewalld                                                                  
-- Logs begin at Mon 2018-12-31 12:16:59 UTC, end at Wed 2019-04-17 09:47:40 UTC. --
Apr 17 09:38:03 systemd[1]: Starting firewalld - dynamic firewall daemon...
Apr 17 09:38:08 systemd[1]: Started firewalld - dynamic firewall daemon.
Apr 17 09:38:09 firewalld[12600]: ERROR: '/usr/bin/iptables-restore -w -n' failed: iptables-restore v1.8.2 (legacy): iptables-restore: unable to initialize table 'security'
                                  
                                  Error occurred at line: 1
                                  Try `iptables-restore -h' or 'iptables-restore --help' for more information.
Apr 17 09:38:09 firewalld[12600]: ERROR: '/usr/bin/ip6tables-restore -w -n' failed: ip6tables-restore v1.8.2 (legacy): ip6tables-restore: unable to initialize table 'security'
                                  
                                  Error occurred at line: 1
                                  Try `ip6tables-restore -h' or 'ip6tables-restore --help' for more information.
Apr 17 09:38:09 firewalld[12600]: ERROR: COMMAND_FAILED: '/usr/bin/ip6tables-restore -w -n' failed: ip6tables-restore v1.8.2 (legacy): ip6tables-restore: unable to initialize table 'security'
                                  
                                  Error occurred at line: 1
                                  Try `ip6tables-restore -h' or 'ip6tables-restore --help' for more information.

```

According to this [discussion in firewalld-users](https://lists.fedorahosted.org/archives/list/firewalld-users@lists.fedorahosted.org/thread/FGFES6CRW5PQVJ36BZK6IEGLA6VRXUWX/), this is not an expected behavior and we should be able to use firewalld without `security` tables.

_Originally posted by @seblu in https://github.com/firewalld/firewalld/issues/411#issuecomment-439737104_

**Comment by erig0** [comment]
@seblu, do you mind trying this patch? File to patch is `/usr/lib/pythonX.Y/site-packages/firewall/core/ipXtables.py`. Replace X.Y with your python version.

```diff
diff --git a/src/firewall/core/ipXtables.py b/src/firewall/core/ipXtables.py                                                                                                
index 68f14a5203db..231aee7eb03d 100644
--- a/src/firewall/core/ipXtables.py
+++ b/src/firewall/core/ipXtables.py
@@ -598,12 +598,13 @@ class ip4tables(object):
     def build_default_rules(self, log_denied="off"):
         default_rules = {}
 
-        default_rules["security"] = [ ]
-        self.our_chains["security"] = set()
-        for chain in BUILT_IN_CHAINS["security"]:
-            default_rules["security"].append("-N %s_direct" % chain)
-            default_rules["security"].append("-A %s -j %s_direct" % (chain, chain))
-            self.our_chains["security"].add("%s_direct" % chain)
+        if self.get_available_tables("security"):
+            default_rules["security"] = [ ]
+            self.our_chains["security"] = set()
+            for chain in BUILT_IN_CHAINS["security"]:
+                default_rules["security"].append("-N %s_direct" % chain)
+                default_rules["security"].append("-A %s -j %s_direct" % (chain, chain))
+                self.our_chains["security"].add("%s_direct" % chain)
 
         default_rules["raw"] = [ ]
         self.our_chains["raw"] = set()
```

**Comment by seblu** [comment]
I'm running version 0.6.3 of firewalld on this server. Looks like `build_default_rules` was massively updated in 9f0eb929c240211d3da978732a9f6930da71486c and I cannot just patch this method without updating the others.

Maybe the simpler solution would to to rebuild a package with upstream git and your patch on top of it.

**Comment by seblu** [comment]
Here is the output of `firewalld --debug --nofork` with the last version from upstream + your patch.
```
2019-04-17 16:08:29 DEBUG1: Loading zone file '/etc/firewalld/zones/drop.xml'
2019-04-17 16:08:29 DEBUG1:   Overloads zone 'drop' ('/usr/lib/firewalld/zones/drop.xml')
2019-04-17 16:08:29 DEBUG1: Loading zone file '/etc/firewalld/zones/public.xml'
2019-04-17 16:08:29 DEBUG1:   Overloads zone 'public' ('/usr/lib/firewalld/zones/public.xml')
2019-04-17 16:08:29 DEBUG1: Loading zone file '/etc/firewalld/zones/trusted.xml'
2019-04-17 16:08:29 DEBUG1:   Overloads zone 'trusted' ('/usr/lib/firewalld/zones/trusted.xml')
2019-04-17 16:08:29 DEBUG1: Using default zone 'public'
2019-04-17 16:08:29 DEBUG1: ipv4 table 'security' does not exist (or not enough permission to check).
2019-04-17 16:08:29 DEBUG1: ipv6 table 'security' does not exist (or not enough permission to check).
2019-04-17 16:08:29 DEBUG1: Flushing rule set
2019-04-17 16:08:29 ERROR: '/usr/bin/iptables-restore -w -n' failed: iptables-restore v1.8.2 (legacy): iptables-restore: unable to initialize table 'security'

Error occurred at line: 1
Try `iptables-restore -h' or 'iptables-restore --help' for more information.

2019-04-17 16:08:29 ERROR: '/usr/bin/ip6tables-restore -w -n' failed: ip6tables-restore v1.8.2 (legacy): ip6tables-restore: unable to initialize table 'security'

Error occurred at line: 1
Try `ip6tables-restore -h' or 'ip6tables-restore --help' for more information.

2019-04-17 16:08:29 DEBUG1: Setting policy to 'ACCEPT'
2019-04-17 16:08:29 ERROR: '/usr/bin/iptables-restore -w -n' failed: iptables-restore v1.8.2 (legacy): iptables-restore: unable to initialize table 'security'

Error occurred at line: 1
Try `iptables-restore -h' or 'iptables-restore --help' for more information.

2019-04-17 16:08:29 ERROR: '/usr/bin/ip6tables-restore -w -n' failed: ip6tables-restore v1.8.2 (legacy): ip6tables-restore: unable to initialize table 'security'

Error occurred at line: 1
Try `ip6tables-restore -h' or 'ip6tables-restore --help' for more information.

2019-04-17 16:08:29 DEBUG1: Traceback (most recent call last):
  File "/usr/lib/python3.7/site-packages/firewall/core/fw.py", line 515, in start
    self._start()
  File "/usr/lib/python3.7/site-packages/firewall/core/fw.py", line 455, in _start
    transaction.execute(True)
  File "/usr/lib/python3.7/site-packages/firewall/core/fw_transaction.py", line 143, in execute
    raise FirewallError(errors.COMMAND_FAILED, errorMsg)
firewall.errors.FirewallError: COMMAND_FAILED: '/usr/bin/ip6tables-restore -w -n' failed: ip6tables-restore v1.8.2 (legacy): ip6tables-restore: unable to initialize table 'security'

Error occurred at line: 1
Try `ip6tables-restore -h' or 'ip6tables-restore --help' for more information.


During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.7/site-packages/firewall/server/decorators.py", line 53, in handle_exceptions
    return func(*args, **kwargs)
  File "/usr/lib/python3.7/site-packages/firewall/server/firewalld.py", line 94, in start
    return self.fw.start()
  File "/usr/lib/python3.7/site-packages/firewall/core/fw.py", line 518, in start
    self.set_policy("ACCEPT")
  File "/usr/lib/python3.7/site-packages/firewall/core/fw.py", line 867, in set_policy
    transaction.execute(True)
  File "/usr/lib/python3.7/site-packages/firewall/core/fw_transaction.py", line 143, in execute
    raise FirewallError(errors.COMMAND_FAILED, errorMsg)
firewall.errors.FirewallError: COMMAND_FAILED: '/usr/bin/ip6tables-restore -w -n' failed: ip6tables-restore v1.8.2 (legacy): ip6tables-restore: unable to initialize table 'security'

Error occurred at line: 1
Try `ip6tables-restore -h' or 'ip6tables-restore --help' for more information.


2019-04-17 16:08:29 ERROR: COMMAND_FAILED: '/usr/bin/ip6tables-restore -w -n' failed: ip6tables-restore v1.8.2 (legacy): ip6tables-restore: unable to initialize table 'security'

Error occurred at line: 1
Try `ip6tables-restore -h' or 'ip6tables-restore --help' for more information.

2019-04-17 16:08:29 DEBUG1: GetAll('org.fedoraproject.FirewallD1')
```

**Comment by seblu** [comment]
If it cans help someone, here is my workaround patch for 0.6.3. Of course, it's ugly.

```diff
--- /usr/lib/python3.7/site-packages/firewall/core/ipXtables.py	2019-03-02 15:39:31.989786288 +0200
+++ /usr/lib/python3.7/site-packages/firewall/core/ipXtables.py	2019-03-02 15:40:42.536633754 +0200
@@ -32,7 +32,6 @@
 import string
 
 BUILT_IN_CHAINS = {
-    "security": [ "INPUT", "OUTPUT", "FORWARD" ],
     "raw": [ "PREROUTING", "OUTPUT" ],
     "mangle": [ "PREROUTING", "POSTROUTING", "INPUT", "OUTPUT", "FORWARD" ],
     "nat": [ "PREROUTING", "POSTROUTING", "OUTPUT" ],
@@ -53,13 +52,6 @@
 LOG_RULES = { }
 OUR_CHAINS = {} # chains created by firewalld
 
-DEFAULT_RULES["security"] = [ ]
-OUR_CHAINS["security"] = set()
-for chain in BUILT_IN_CHAINS["security"]:
-    DEFAULT_RULES["security"].append("-N %s_direct" % chain)
-    DEFAULT_RULES["security"].append("-I %s 1 -j %s_direct" % (chain, chain))
-    OUR_CHAINS["security"].add("%s_direct" % chain)
-
 DEFAULT_RULES["raw"] = [ ]
 OUR_CHAINS["raw"] = set()
 for chain in BUILT_IN_CHAINS["raw"]:
```

**Comment by erig0** [comment]
@seblu, I missed a couple spots in the patch against master. Try this patch:

```diff
diff --git a/src/firewall/core/ipXtables.py b/src/firewall/core/ipXtables.py                                                                                                
index 68f14a5203db..2b3d6e4c58be 100644
--- a/src/firewall/core/ipXtables.py
+++ b/src/firewall/core/ipXtables.py
@@ -541,6 +541,8 @@ class ip4tables(object):
         self.rich_rule_priority_counts = {}
         rules = []
         for table in BUILT_IN_CHAINS.keys():
+            if not self.get_available_tables(table):
+                continue
             # Flush firewall rules: -F
             # Delete firewall chains: -X
             # Set counter to zero: -Z
@@ -552,6 +554,8 @@ class ip4tables(object):
         rules = []
         _policy = "DROP" if policy == "PANIC" else policy
         for table in BUILT_IN_CHAINS.keys():
+            if not self.get_available_tables(table):
+                continue
             if table == "nat":
                 continue
             for chain in BUILT_IN_CHAINS[table]:
@@ -598,12 +602,13 @@ class ip4tables(object):
     def build_default_rules(self, log_denied="off"):
         default_rules = {}
 
-        default_rules["security"] = [ ]
-        self.our_chains["security"] = set()
-        for chain in BUILT_IN_CHAINS["security"]:
-            default_rules["security"].append("-N %s_direct" % chain)
-            default_rules["security"].append("-A %s -j %s_direct" % (chain, chain))
-            self.our_chains["security"].add("%s_direct" % chain)
+        if self.get_available_tables("security"):
+            default_rules["security"] = [ ]
+            self.our_chains["security"] = set()
+            for chain in BUILT_IN_CHAINS["security"]:
+                default_rules["security"].append("-N %s_direct" % chain)
+                default_rules["security"].append("-A %s -j %s_direct" % (chain, chain))
+                self.our_chains["security"].add("%s_direct" % chain)
 
         default_rules["raw"] = [ ]
         self.our_chains["raw"] = set()
```

**Comment by seblu** [comment]
Looks good. No more errors. Rules are loaded!

Thanks.

**Comment by erig0** [comment]
> Looks good. No more errors. Rules are loaded!

Thanks for confirming. I'll push this fix soon. I want to take a look at the other tables first.

**Comment by Francewhoa** [comment]
For those facing this challenge, it is fixed in FirewallD [version 0.6.4 or more recent](https://github.com/firewalld/firewalld/releases)

>This is a bug fix only release:
>fix: ipXtables: don't use tables that aren't available
